File: mobile/src/api/barbers.ts
import { apiClient } from "./client";
import { ApiError } from "./client";
import type { AvailableSlot, Barber, BarberFormPayload, BarberService } from "../types/barber";

const fallbackImage = "https://images.unsplash.com/photo-1621605815971-fbc98d665033?auto=format&fit=crop&w=300&q=80";

type RawBarber = Partial<Barber> & {
  name?: string | null;
  experience_years?: number | null;
};

type RawSlot =
  | string
  | {
      time?: string | null;
      start_time?: string | null;
      available?: boolean | null;
      is_available?: boolean | null;
      is_booked?: boolean | null;
      is_expired?: boolean | null;
      reason?: string | null;
    };

export type BarberSort = "nearest" | "cheapest" | "expensive";

export interface PublicBarberParams {
  sort?: BarberSort;
  userLat?: number;
  userLng?: number;
}

function asList<T>(value: unknown): T[] {
  if (Array.isArray(value)) {
    return value as T[];
  }
  if (value && typeof value === "object") {
    const objectValue = value as { items?: unknown; slots?: unknown; data?: unknown };
    if (Array.isArray(objectValue.items)) {
      return objectValue.items as T[];
    }
    if (Array.isArray(objectValue.slots)) {
      return objectValue.slots as T[];
    }
    if (Array.isArray(objectValue.data)) {
      return objectValue.data as T[];
    }
  }
  return [];
}

export function normalizeBarber(barber: RawBarber): Barber {
  const id = Number(barber.id);
  const fullName = barber.full_name || barber.name || "Barber";
  const specialty = barber.specialty || "Professional Barber";
  const rating = barber.average_rating ?? barber.rating ?? 0;
  const photoUrl = barber.avatar || barber.photo_url || fallbackImage;
  const yearsExperience = barber.years_experience ?? barber.experience_years ?? null;

  return {
    ...barber,
    id,
    user_id: barber.user_id ?? id,
    email: barber.email ?? "",
    full_name: fullName,
    specialty,
    barbershop_name: barber.barbershop_name ?? null,
    photo_url: photoUrl,
    avatar: photoUrl,
    phone: barber.phone ?? null,
    rating,
    average_rating: rating,
    years_experience: yearsExperience,
    price_from: barber.price_from ?? null,
    base_price: barber.base_price ?? barber.price_from ?? null,
    latitude: barber.latitude ?? null,
    longitude: barber.longitude ?? null,
    address: barber.address ?? null,
    work_start_time: barber.work_start_time ?? null,
    work_end_time: barber.work_end_time ?? null,
    off_days: barber.off_days ?? [],
    distance_km: barber.distance_km ?? null,
    is_active: barber.is_active ?? true,
    created_at: barber.created_at ?? "",
    updated_at: barber.updated_at ?? "",
  };
}

function normalizeSlot(slot: RawSlot): AvailableSlot {
  if (typeof slot === "string") {
    return {
      time: slot.slice(0, 5),
      is_available: true,
      is_booked: false,
      is_expired: false,
    };
  }

  const time = slot.time || slot.start_time || "";

  const isBooked = slot.is_booked ?? (slot.reason === "booked");
  const isExpired = slot.is_expired ?? (slot.reason === "expired");
  const isAvailable = slot.is_available ?? slot.available ?? !(isBooked || isExpired);

  return {
    time: time.slice(0, 5),
    is_available: isAvailable,
    is_booked: isBooked,
    is_expired: isExpired,
    available: slot.available ?? isAvailable,
    reason: slot.reason ?? null,
  };
}

export async function getPublicBarbers(params: PublicBarberParams = {}): Promise<Barber[]> {
  const query: Record<string, string | number> = {};
  if (params.sort) query.sort = params.sort;
  if (typeof params.userLat === "number") query.user_lat = params.userLat;
  if (typeof params.userLng === "number") query.user_lng = params.userLng;
  const response = await apiClient.get<unknown>("/public/barbers", { params: query });
  return asList<RawBarber>(response.data).map(normalizeBarber).filter((barber) => Number.isFinite(barber.id));
}

export async function getPublicBarber(id: number): Promise<Barber> {
  const response = await apiClient.get<RawBarber>(`/public/barbers/${id}`);
  return normalizeBarber(response.data);
}

export async function getPublicBarberServices(barberId: number): Promise<BarberService[]> {
  const response = await apiClient.get<unknown>(`/public/barbers/${barberId}/services`);
  return asList<BarberService>(response.data);
}

export async function getAvailableSlots(barberId: number, date: string, serviceId?: number): Promise<AvailableSlot[]> {
  const params = { date, service_id: serviceId };
  try {
    const response = await apiClient.get<unknown>(`/public/barbers/${barberId}/available-slots`, {
      params,
      silentStatuses: [404],
    });
    return asList<RawSlot>(response.data).map(normalizeSlot).filter((slot) => slot.time);
  } catch (error) {
    if (!(error instanceof ApiError) || error.status !== 404) {
      throw error;
    }
    const response = await apiClient.get<unknown>(`/public/barbers/${barberId}/availability`, { params });
    return asList<RawSlot>(response.data).map(normalizeSlot).filter((slot) => slot.time);
  }
}

export async function getAdminBarbers(): Promise<Barber[]> {
  const response = await apiClient.get<Barber[]>("/barbers");
  return response.data;
}

export async function createAdminBarber(payload: BarberFormPayload): Promise<Barber> {
  const response = await apiClient.post<Barber>("/barbers", payload);
  return response.data;
}

File: backend/app/api/routes/barbers.py
from datetime import date
from math import asin, cos, radians, sin, sqrt
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.orm import Session, selectinload

from app.core.config import settings
from app.db.database import get_db
from app.models.barber import Barber
from app.models.barber_service import BarberService
from app.schemas.barber import AvailableSlot, AvailableSlotsResponse, BarberRead, BarberServiceRead
from app.services.booking_service import get_available_slots, get_available_slots_response

router = APIRouter()


def haversine_km(user_lat: float, user_lng: float, barber: Barber) -> float | None:
    if barber.latitude is None or barber.longitude is None:
        return None

    earth_radius_km = 6371.0
    lat1, lng1, lat2, lng2 = map(radians, [user_lat, user_lng, barber.latitude, barber.longitude])
    d_lat = lat2 - lat1
    d_lng = lng2 - lng1
    value = sin(d_lat / 2) ** 2 + cos(lat1) * cos(lat2) * sin(d_lng / 2) ** 2
    return round(2 * earth_radius_km * asin(sqrt(value)), 1)


def serialize_barber(barber: Barber, distance_km: float | None = None) -> BarberRead:
    return BarberRead(
        id=barber.id,
        user_id=barber.user_id,
        email=barber.user.email,
        full_name=barber.full_name,
        specialty=barber.specialty,
        barbershop_name=barber.barbershop_name,
        photo_url=barber.photo_url,
        phone=barber.phone,
        rating=barber.rating,
        years_experience=barber.years_experience,
        price_from=barber.price_from if barber.price_from is not None else barber.base_price,
        base_price=barber.base_price if barber.base_price is not None else barber.price_from,
        latitude=barber.latitude,
        longitude=barber.longitude,
        address=barber.address,
        work_start_time=barber.work_start_time,
        work_end_time=barber.work_end_time,
        off_days=barber.off_days or [],
        bio=barber.bio,
        is_active=barber.is_active,
        is_financially_blocked=barber.is_financially_blocked,
        created_at=barber.created_at,
        updated_at=barber.updated_at,
        distance_km=distance_km,
        experience_years=barber.years_experience,
    )


@router.get("/barbers", response_model=list[BarberRead])
def list_public_barbers(
    sort: Literal["nearest", "cheapest", "expensive"] | None = None,
    user_lat: float | None = None,
    user_lng: float | None = None,
    db: Session = Depends(get_db),
) -> list[BarberRead]:
    barbers = db.scalars(
        select(Barber)
        .options(selectinload(Barber.user))
        .where(Barber.is_active.is_(True))
    ).all()
    if settings.financial_blocking_enabled:
        barbers = [barber for barber in barbers if not barber.is_financially_blocked]

    distances: dict[int, float | None] = {}
    if user_lat is not None and user_lng is not None:
        distances = {barber.id: haversine_km(user_lat, user_lng, barber) for barber in barbers}

    if sort == "nearest":
        if user_lat is None or user_lng is None:
            raise HTTPException(status_code=400, detail="Nearest sort requires user_lat and user_lng")
        barbers = sorted(barbers, key=lambda barber: (distances[barber.id] is None, distances[barber.id] or 0, barber.full_name))
    elif sort == "cheapest":
        barbers = sorted(barbers, key=lambda barber: ((barber.price_from if barber.price_from is not None else barber.base_price) is None, barber.price_from if barber.price_from is not None else barber.base_price or 0, barber.full_name))
    elif sort == "expensive":
        barbers = sorted(barbers, key=lambda barber: ((barber.price_from if barber.price_from is not None else barber.base_price) is None, -((barber.price_from if barber.price_from is not None else barber.base_price) or 0), barber.full_name))
    else:
        barbers = sorted(barbers, key=lambda barber: barber.full_name)

    return [serialize_barber(barber, distances.get(barber.id)) for barber in barbers]


@router.get("/barbers/{barber_id}", response_model=BarberRead)
def get_public_barber(barber_id: int, db: Session = Depends(get_db)) -> BarberRead:
    barber = db.scalar(
        select(Barber)
        .options(selectinload(Barber.user))
        .where(Barber.id == barber_id, Barber.is_active.is_(True))
    )
    if not barber or (settings.financial_blocking_enabled and barber.is_financially_blocked):
        raise HTTPException(status_code=404, detail="Barber not found")
    return serialize_barber(barber)


@router.get("/barbers/{barber_id}/services", response_model=list[BarberServiceRead])
def public_barber_services(barber_id: int, db: Session = Depends(get_db)) -> list[BarberService]:
    barber = db.get(Barber, barber_id)
    if not barber or not barber.is_active or (settings.financial_blocking_enabled and barber.is_financially_blocked):
        raise HTTPException(status_code=404, detail="Barber not found")
    return list(
        db.scalars(
            select(BarberService)
            .where(BarberService.barber_id == barber_id, BarberService.is_active.is_(True))
            .order_by(BarberService.price, BarberService.name)
        ).all()
    )


@router.get("/barbers/{barber_id}/availability", response_model=list[AvailableSlot])
def available_slots(
    barber_id: int,
    date: date,
    service_id: int | None = None,
    db: Session = Depends(get_db),
) -> list[AvailableSlot]:
    return get_available_slots(db, barber_id, date, service_id)


@router.get("/barbers/{barber_id}/available-slots", response_model=AvailableSlotsResponse)
def available_slots_v2(
    barber_id: int,
    date: date,
    service_id: int | None = None,
    db: Session = Depends(get_db),
) -> AvailableSlotsResponse:
    return get_available_slots_response(db, barber_id, date, service_id)
